include/circt/Dialect/Seq/Seq.td

```diff
@@ -61,4 +61,47 @@ def CompRegOp : SeqOp<"compreg",
   ];
 }
 
+def FirRegOp : SeqOp<"firreg",
+    [NoSideEffect, AllTypesMatch<["input", "data"/*, "resetValue"*/]>,
+     SameVariadicOperandSize,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+      // AllTypesMatch doesn't work with Optional types yet.
+
+  let summary = "Register with preset and sync or async reset";
+  let description = [{
+    See the Seq dialect rationale for a longer description.
+
+    ```
+    %name = seq.firreg %input, %clk [ sym @sym]
```

Comment:
The ironic thing about data-flow input is this op then doesn't work generally in SSA regions, which likely are what we want for any analyzable middle IR.

Reply:
I think we'll want two separate things for this: an HW register (or `RegLike` with different impls by reg flavor) for `Graph` regions, and something like a memory-load-store register for a middle IR that wants def-before-use.

I'd argue the Seq dialect should form a `HW+Comb+Seq` group with the same flavor of representation (graph regions), and this `seq.firreg` falls into that category.

Reply:
This register is specifically for what we lower FIRRTL to, not designed to replace the registers at the FIRRTL level.